File: python/QEasyWidgets/QEasyWidgets/Components/Browser.py
from pathlib import Path
from typing import Optional, overload
from PySide6.QtGui import *
from PySide6.QtCore import *
from PySide6.QtWidgets import *

from ..Common.StyleSheet import *
from ..Common.QFunctions import *
from ..Resources.Sources import *
from .Widget import SizableWidget
from .ScrollArea import ScrollDelegate
from .Menu import MenuBase

##############################################################################################################################

class TextBrowserBase(QTextBrowser, SizableWidget):
    """
    Base class for textBrowser components
    """
    def __init__(self, parent: Optional[QWidget] = None):
        super().__init__(parent)

        self.scrollDelegate = ScrollDelegate(self)

        StyleSheetBase.Browser.apply(self)

    def contextMenuEvent(self, e):
        menu = MenuBase(self)
        menu.exec(e.globalPos())

    def loadMarkdown(self, file: Union[str, Path]):
        with open(file, mode = 'r', encoding = 'utf-8') as f:
            md = f.read()
        self.setMarkdown(md)

    def loadHtml(self, file: Union[str, Path]):
        with open(file, mode = 'r', encoding = 'utf-8') as f:
            html = f.read()
        self.setHtml(html)

    def setBorderless(self, borderless: bool) -> None:
        self.setProperty("isBorderless", borderless)

    def clearDefaultStyleSheet(self) -> None:
        StyleSheetBase.Browser.deregistrate(self)

##############################################################################################################################

File: c++/QEasyWidgets/Components/Browser.cpp
#include "Browser.h"

#include <QFile>
#include <QTextStream>
#include <QContextMenuEvent>

#include "../Common/StyleSheet.h"
#include "Menu.h"


/**
 * TextBrowserBase implementation
 */

TextBrowserBase::TextBrowserBase(QWidget *parent)
    : SizableWidget<QTextBrowser>(parent) {
    init();
}

void TextBrowserBase::init() {
    StyleSheetBase::apply(this, StyleSheetBase::Browser);
}

void TextBrowserBase::loadMarkdown(const QString &filePath) {
    QFile file(filePath);
    if (file.open(QIODevice::ReadOnly | QIODevice::Text)) {
        QTextStream in(&file);
        in.setEncoding(QStringConverter::Utf8);
        QString markdown = in.readAll();
        setMarkdown(markdown);
        file.close();
    }
}

void TextBrowserBase::loadHtml(const QString &filePath) {
    QFile file(filePath);
    if (file.open(QIODevice::ReadOnly | QIODevice::Text)) {
        QTextStream in(&file);
        in.setEncoding(QStringConverter::Utf8);
        QString html = in.readAll();
        setHtml(html);
        file.close();
    }
}

void TextBrowserBase::setBorderless(bool borderless) {
    setProperty("isBorderless", borderless);
}

void TextBrowserBase::clearDefaultStyleSheet() {
    StyleSheetBase::deregistrate(this);
}

void TextBrowserBase::contextMenuEvent(QContextMenuEvent *event) {
    MenuBase menu(this);
    menu.exec(event->globalPos());
}